### ISSUE TITLE
sample filters and lua scripts needed

### ISSUE THREAD

**Issue by vj777** [mpcjanssen/simpletask-android#737]
The initial install gives no preset filters nor LuaScripts. The help file does not give any info or samples on how to set up these essential items. 
Is there somewhere that existing luascripts can be obtained from? Could that be referenced/linked from the Help?
This would be simple to provide but would REALLY help users. 
the help should describe how applying filters & scripts affect each other eg. is the effect of scripts cumulative?
Thanks

**Comment by mpcjanssen** [comment]
The using Lua help has some examples. The actual way the filtering works is not described in detail but is not that complex. There is a global Lua code piece. And Lua code which is part of a filter. The app calls specific callbacks where functions in the filter win.

**Comment by likwrk** [comment]
I agree with @vj777 
I tried to write some scripts:
1. Tasks without tags.
```
function onFilter (task, f, ext)
  return #f.tags < 0
end
```

2. Task for today
```
t = os.date("*t")
start_today = os.time{year=t.year, month=t.month, day=t.day, hour=0}
end_today = start_today + 24 * 60 * 60
function onFilter (task, f, ext)
  return f.due and f.due > start_today and f.due < end_today
end
```

And both returned no task

Writing scripts help really needed. Thanks!

**Comment by mpcjanssen** [comment]
Writing help is a lot of work. One of the reason I picked Lua instead of inventing some other language myself is that you can find help online easily.  So much more scripting help is not going to happen any time soon.

**Comment by smichel17** [comment]
I've been thinking about making it so instead of defining all your callbacks in that one script box, you pick a function to define using the drop down menu and write them one at a time (ie, ST would wrap whatever you wrote with `function onFilter(t,f,e)` and `end`

The reason for this is forcing everybody to use the same variable names for their arguments. This would allow anybody to copy and paste to use other people's scripts. Then we could make an issue here just for people to share scripts with one another and link it from within the app.

On the downside, this would also be quite a bit of work, would break existing scripts and force people to update, and would be rather nonstandard, so this particular behavior would require documentation.

**Comment by mpcjanssen** [comment]
@smichel17 those are some considerable downsides. Sharing scripts is still easy if you include the function definition (which you should). So variable names is not really an issue IMO.

PRs against introducing sample scripts are welcome though. It would be nice to have a samples script help page.

**Comment by smichel17** [comment]
@mpcjanssen yeah. I would not go ahead and work on that particular idea unless I thought of a way to reduce the downsides. I have still been thinking of it, though.

As for example scripts, I really prefer inline help to separate documentation. It's harder to build, but much more useful. In this case, I think I'd rather ship with default saved filters, some of which include scripts. Now people have example scripts and can play around with them.

**Comment by DDT3000** [comment]
Not sure if this an appropriate place to ask questions, but as an example, I am looking to write a script that only shows groups (when grouping by tags) that contain at least one item with priority (A) - having one such task per project/tag is my way of representing active GTD projects. At the moment, I am not sure what is the right way to do this - via onFilter, via onGroup, or some other way, or maybe it cannot be done - does not seem like onGroup has access to other items in the group (which makes sense, since that depends on return values of onGroup for all other tasks).
Having more example scripts would maybe steer me in the right direction instead of indirectly asking a question here. (this is perhaps one last thing I cannot figure out how do in simpletask, which has been wonderful otherwise).

**Comment by mpcjanssen** [comment]
@DDT3000 something like that will require having acces to a full copy of the todolist while calling the `onFilter` callback. Currently that is not supported.

For your use case you could consider having a naming convention for inactive projects (like `+project_archive` then it would be easy to filter out inactive projects when grouping by tag. By checking if the project name matches the inactive naming convention.

**Comment by DDT3000** [comment]
@mpcjanssen Does onFilter have access to variables in global scope, or something to that effect? The one thing I could rely on is the fact that tasks with defined priority are always on the top of my list, so as long as onFilter is called in that order, I could maybe maintain that global list of tags that correlate to active projects, and update/access it from onFilter.
Thanks for the other suggestion, my process so far involves changing priority of projects on a weekly basis, so retagging each task every time this happens would be problematic.  My other idea now is to generate a list of these tags externally, as a pre-saved filter, and upload to phone every week (or even run a python script on the phone itself).

**Comment by mpcjanssen** [comment]
@DDT3000 all code is running in a single Lua interpreter within different `modules`. You can use global vars and they will be retained between `onDisplay` calls. What is missing to make good use of this is callbacks when the filtering starts to initiliaze these vars.
